Firestore/Source/API/FSTUserDataReader.mm

```diff
@@ -277,51 +318,70 @@ - (FieldValue)parsedQueryValue:(id)input allowArrays:(bool)allowArrays {
       if (context.array_element() && context.data_source() != UserDataSource::ArrayArgument) {
         ThrowInvalidArgument("Nested arrays are not supported");
       }
-      return [self parseArray:(NSArray *)input context:std::move(context)];
+      return absl::optional<google_firestore_v1_Value>([self parseArray:(NSArray *)input
+                                                                context:std::move(context)]);
     } else {
-      return [self parseScalarValue:input context:std::move(context)];
+      return absl::optional<google_firestore_v1_Value>([self parseScalarValue:input
+                                                                      context:std::move(context)]);
     }
   }
 }
 
-- (FieldValue)parseDictionary:(NSDictionary<NSString *, id> *)dict
-                      context:(ParseContext &&)context {
+- (google_firestore_v1_Value)parseDictionary:(NSDictionary<NSString *, id> *)dict
+                                     context:(ParseContext &&)context {
+  google_firestore_v1_Value result;
+  result.which_value_type = google_firestore_v1_Value_map_value_tag;
+
+  __block std::vector<google_firestore_v1_MapValue_FieldsEntry> fields;
+
   if (dict.count == 0) {
     const FieldPath *path = context.path();
     if (path && !path->empty()) {
       context.AddToFieldMask(*path);
     }
-    return ObjectValue::Empty().AsFieldValue();
   } else {
-    __block ObjectValue result = ObjectValue::Empty();
-
+    // Populate a vector of fields since we don't know the size of the final fields array.
     [dict enumerateKeysAndObjectsUsingBlock:^(NSString *key, id value, BOOL *) {
-      absl::optional<FieldValue> parsedValue =
+      absl::optional<google_firestore_v1_Value> parsedValue =
```

Comment:
Question: what does it take to know whether the value can be parsed? What I'm getting at is whether it could be feasible to do a first pass on `dict` to get the array size, then create the values right in the array and skip the vector.

Reply:
We have to check whether it is a FIeldValue sentinel, which would duplicate the existing check in `parseData`. I don't think we can remove the `parseData` check since this function has more than this callsite.

Reply:
Ok, if I'm reading the code correctly, it amounts to going through the dictionary and calling `[input isKindOfClass:[FIRFieldValue class]]` on each element. If this is correct, then I strongly suspect that it's faster than creating and filling a vector and then copying vector elements into the array. What do you think about doing this, perhaps in a follow-up?

Reply:
I updated this in this PR.